src/nm.c

```diff
@@ -962,49 +919,108 @@ netplan_netdef_write_nm(
     return no_error;
 }
 
-static void
-nd_append_non_nm_ids(gpointer data, gpointer str)
-{
-    const NetplanNetDefinition* nd = data;
-
-    if (nd->backend != NETPLAN_BACKEND_NM) {
-        if (nd->match.driver) {
-            /* TODO: NetworkManager supports (non-globbing) "driver:..." matching nowadays */
-            /* NM cannot match on drivers, so ignore these via udev rules */
-            if (!udev_rules)
-                udev_rules = g_string_new(NULL);
-            g_string_append_printf(udev_rules, "ACTION==\"add|change\", SUBSYSTEM==\"net\", ENV{ID_NET_DRIVER}==\"%s\", ENV{NM_UNMANAGED}=\"1\"\n", nd->match.driver);
-        } else {
-            g_string_append_netdef_match((GString*) str, nd);
-        }
-    }
-}
-
 gboolean
 netplan_state_finish_nm_write(
         const NetplanState* np_state,
         const char* rootdir,
         GError** error)
 {
-    GString *s = NULL;
-    gsize len;
+    GString* udev_rules = g_string_new(NULL);
+    GString *nm_conf = g_string_new(NULL);
 
     if (netplan_state_get_netdefs_size(np_state) == 0)
         return TRUE; // LCOV_EXCL_LINE as generate.c already deals with it.
 
     /* Set all devices not managed by us to unmanaged, so that NM does not
-     * auto-connect and interferes */
-    s = g_string_new("[keyfile]\n# devices managed by networkd\nunmanaged-devices+=");
-    len = s->len;
-    g_list_foreach(np_state->netdefs_ordered, nd_append_non_nm_ids, s);
-    if (s->len > len)
-        g_string_free_to_file(s, rootdir, "run/NetworkManager/conf.d/netplan.conf", NULL);
+     * auto-connect and interferes.
+     * Also, mark all devices managed by us explicitly, so it won't get in
+     * conflict with the system's udev rules that might ignore some devices
+     * in containers via usr/lib/udev/rules.d/85-nm-unmanaged-devices.rules */
+    GList* iter = np_state->netdefs_ordered;
```

Comment:
praise: great work on the whole function body, very readable and well commented despite the complex logic!